### DISCUSSION TITLE
What code can you run inside effect transactions?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#60] [Q&A] [answered]
When running a transaction, can I run normal code? I mean code that can conditionally call a setter function.

If no setter is called, will the transaction trigger a rebuild?

**Comment by GregoryConrad** [comment, marked as answer]
> I mean code that can conditionally call a setter function.

Yes. Don't run anything that can throw an exception though--that is not supported ATM.

`container.read`, `container.runTransaction`, and side effect rebuilds are all fair game. Just be a little careful that the data that you `container.read` is _actually_  immutable; i.e., imagine the following:
```dart
container.runTransaction(() {
  final (_, setState) = container.read(someStateGetterSetterCapsule);
  setState(1234);
  final (getState, _) = container.read(someStateGetterSetterCapsule);
  // invoking getState here will return 1234, despite the rebuild not happening until later!
);
```
This is another one of those issues that isn't an issue in Rust. It's also a symptom of how Dart works and how I designed the API around that. I suppose I might be able to do a fix where you pass any side effect state mutations _inside_ of a call to rebuild like is done in Rust without any breaking changes--that is an idea that might be worth exploring. Edit: #61 

> If no setter is called, will the transaction trigger a rebuild?

No, only capsules that triggered a rebuild during the transaction will actually rebuild.

**Reply by GregoryConrad** [reply]
@busslina I just fixed the above problem with getter/setters in #64. It should be released soon!
